src/Microsoft.VisualStudio.Editor.Razor/DefaultVisualStudioRazorParser.cs

```diff
@@ -0,0 +1,372 @@
+            string filePath,
+            string projectDirectory,
+            ErrorReporter errorReporter,
+            ICompletionBroker completionBroker,
+            IEditorOperationsFactoryService editorOperationsFactory,
+            IEnumerable<ICanHasContextChangedListener> parserContextChangedListeners)
+        {
+            if (dispatcher == null)
+            {
+                throw new ArgumentNullException(nameof(dispatcher));
+            }
+
+            if (documentTracker == null)
+            {
+                throw new ArgumentNullException(nameof(documentTracker));
+            }
+
+            if (templateEngineFactory == null)
+            {
+                throw new ArgumentNullException(nameof(templateEngineFactory));
+            }
+
+            if (string.IsNullOrEmpty(filePath))
+            {
+                throw new ArgumentException(Resources.ArgumentCannotBeNullOrEmpty, nameof(filePath));
+            }
+
+            if (string.IsNullOrEmpty(projectDirectory))
+            {
+                throw new ArgumentException(Resources.ArgumentCannotBeNullOrEmpty, nameof(projectDirectory));
+            }
+
+            if (errorReporter == null)
+            {
+                throw new ArgumentNullException(nameof(errorReporter));
+            }
+
+            if (completionBroker == null)
+            {
+                throw new ArgumentNullException(nameof(completionBroker));
+            }
+
+            if (editorOperationsFactory == null)
+            {
+                throw new ArgumentNullException(nameof(editorOperationsFactory));
+            }
+
+            if (parserContextChangedListeners == null)
+            {
+                throw new ArgumentNullException(nameof(parserContextChangedListeners));
+            }
+
+            _dispatcher = dispatcher;
+            _templateEngineFactory = templateEngineFactory;
+            _errorReporter = errorReporter;
+            _completionBroker = completionBroker;
+            _parserContextChangedListeners = parserContextChangedListeners;
+            _documentTracker = documentTracker;
+            FilePath = filePath;
+            _projectDirectory = projectDirectory;
+            TextBuffer = _documentTracker.TextBuffer;
+
+            _documentTracker.ContextChanged += DocumentTracker_ContextChanged;
+            TextBuffer.Changed += TextBuffer_OnChanged;
+            _braceSmartIndenter = new BraceSmartIndenter(_dispatcher, _documentTracker, editorOperationsFactory);
+        }
+
+        public override event EventHandler<DocumentStructureChangedEventArgs> DocumentStructureChanged;
+
+        public override RazorTemplateEngine TemplateEngine => _templateEngine;
+
+        public override string FilePath { get; }
+
+        public override RazorCodeDocument CodeDocument => _codeDocument;
+
+        public override ITextSnapshot Snapshot => _snapshot;
+
+        public override ITextBuffer TextBuffer { get; }
```

Comment:
nit: I don't super-love mixing field-backed properties and auto-props. It makes it a bit harder to chase down what's going the on.